In [33]:
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
import numpy as np

In [34]:
products_data=pd.read_csv('balancedDATA.csv')
products_data

NameDescription  \
0      surco a conversion bar bike rack featuresused ...   
1      kent marine  turbo calcium ounce bottle turbo ...   
2      inch pro style swim snorkel breathe underwater...   
3      zia natural skincare seaweed lift serum ounce ...   
4      nexcare waterproof bandage knee amp elbow  ct ...   
...                                                  ...   
35761  palladio rice paper natural unique tissue made...   
35762  learning resource tool set piece molded plasti...   
35763  serious skin care aeye xr nanoencapsulated eye...   
35764  wireless phone xcessories travel charger samsu...   
35765  fsgt g ch gun transmitter radio controlled car...   

                      categories  
0                     Automotive  
1                   Pet Supplies  
2              Sports & Outdoors  
3                         Beauty  
4         Health & Personal Care  
...                          ...  
35761      Arts, Crafts & Sewing  
35762               Toys & Games  
35763                     Beauty  
35764  Cell Phones & Accessories  
35765               Toys & Games  

[35766 rows x 2 columns]

In [35]:
products_data.categories.value_counts()

Automotive                   1987
Pet Supplies                 1987
Grocery & Gourmet Food       1987
Office Products              1987
Industrial & Scientific      1987
Tools & Home Improvement     1987
Electronics                  1987
Musical Instruments          1987
Appliances                   1987
Clothing, Shoes & Jewelry    1987
Baby Products                1987
Toys & Games                 1987
Cell Phones & Accessories    1987
Arts, Crafts & Sewing        1987
Health & Personal Care       1987
Beauty                       1987
Sports & Outdoors            1987
Patio, Lawn & Garden         1987
Name: categories, dtype: int64

# Rendre label numerique

In [36]:
l=[]
for text in products_data["categories"].unique():
    l.append(text)

y=[]
for text in products_data["categories"]:
    for j in range(18):
        if text==l[j]:
            y.append(j)
    
y = np.asarray(y)

yy = pd.Series(y)
yy

0        0
1        1
2        2
3        3
4        4
        ..
35761    5
35762    7
35763    3
35764    6
35765    7
Length: 35766, dtype: int64

# Extracting Features From TextData using TF-IDF

In [37]:
from sklearn import feature_extraction, model_selection, naive_bayes, pipeline, manifold, preprocessing

In [38]:
Text=[]
for text in products_data["NameDescription"]:
    Text.append(text)

In [39]:
from sklearn.feature_extraction.text import TfidfVectorizer
# First, we need a vectorizer object
vectorizer = TfidfVectorizer()
x = vectorizer.fit_transform(Text).toarray()
xx=  pd.DataFrame(x)
xx

0      1      2      3      4      5      6      7      8      9      \
0        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4        0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...      ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
35761    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35762    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35763    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35764    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35765    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       ...  99060  99061  99062  99063  99064  99065  99066  99067  99068  \
0      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
1      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
2      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
3      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
4      ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
...    ...    ...    ...    ...    ...    ...    ...    ...    ...    ...   
35761  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35762  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35763  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35764  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   
35765  ...    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0    0.0   

       99069  
0        0.0  
1        0.0  
2        0.0  
3        0.0  
4        0.0  
...      ...  
35761    0.0  
35762    0.0  
35763    0.0  
35764    0.0  
35765    0.0  

[35766 rows x 99070 columns]

After feature extraction from our text it gives 99070 features.

# MLP befor feature selection

# 10 iteration

In [40]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,random_state=1,test_size=0.2)

In [41]:
#max_iter=500
from sklearn.neural_network import MLPClassifier
classifier = MLPClassifier(hidden_layer_sizes=(250,100,50), max_iter=10, warm_start=True, verbose=True)
classifier.fit(x_train,y_train)

Iteration 1, loss = 1.57418563
Iteration 2, loss = 0.27196261
Iteration 3, loss = 0.06753563
Iteration 4, loss = 0.02881509
Iteration 5, loss = 0.01949127
Iteration 6, loss = 0.01447590
Iteration 7, loss = 0.01277290
Iteration 8, loss = 0.01161739
Iteration 9, loss = 0.01185376
Iteration 10, loss = 0.00974388


MLPClassifier(hidden_layer_sizes=(250, 100, 50), max_iter=10, verbose=True,
              warm_start=True)

In [42]:
y_pred = classifier.predict(x_test)

In [43]:
from sklearn.metrics import accuracy_score
print("Number of features is", x_train.shape[1])
print("Test set accuracy of MLPClassifier Classifier is {}".format(round(accuracy_score(y_pred,y_test)*100,2)))

Number of features is 99070
Test set accuracy of MLPClassifier Classifier is 85.67


In [44]:
# save model as pickle file

In [45]:
import pickle
pickle.dump(classifier, open('18--10last.pkl', 'wb'))

In [46]:
loaded_model = pickle.load(open('18--10last.pkl', 'rb'))
result = loaded_model.score(x_test, y_test)
print(result)

0.8567235113223372
